### PR TITLE
runtime/logging: only suppress payloads for handlers that compress responses

### PR DESCRIPTION
**Pull request by srenatus**
To get compressed responses, two things need to be true:

1. The client must accept compressed responses
2. The handler must reply with a compressed response

For general API requests, (1) holds most of the time. (2) is only true
for the metrics endpoint at them moment, since the 3rd party library we
use for serving the prometheus endpoint will do compression.

### CODE REVIEW

File: topdown/eval.go [hunk 1]
```diff
@@ -664,18 +664,12 @@ func (e *eval) evalNotPartialSupport(negationID uint64, expr *ast.Expr, unknowns
 	}
 
 	// Save expression that refers to support rule set.
-
-	terms := expr.Terms
-	expr.Terms = nil // Prevent unnecessary copying the terms.
-	cpy := expr.Copy()
-	expr.Terms = terms
+	cpy := copyExprSansTerms(expr)
```
Comment (ashutosh-narkar) [inline topdown/eval.go:667]: Is this change related to the payload suppress change ?

**Comment by ashutosh-narkar** [comment]
Is there a way to see the output difference between the existing and new behavior ?

File: topdown/eval.go (hunk 1 above)
Comment (srenatus) [reply topdown/eval.go:667]: 🙈 this is where it went. I had been looking for that. It's #4503. Will fix this PR.

**Comment by srenatus** [comment]
> Is there a way to see the output difference between the existing and new behavior ?

Currently no logs are ever included if the client accepts gzip. So we're fixing a bug 😄 I can look into adding a test tomorrow 👍

**Review by anderseknert — APPROVED** [review]
LGTM! Quite some tests!
